Makefile pr

Djfe opened the pull request:
so yeah, I pretty much reworked the whole build system 😅
if you got questions, leave them here

Djfe committed:
site.conf: move comments around, fix indentation

remove show_altitude, it's default false anyways
remove community-build mirror (experimental channel)

Djfe committed:
site.conf: add new signing key: build server

Djfe committed:
site.conf: remove weather radar channels from outdoor_chanlist

Djfe committed:
site.mk: rework packagelist variables

sort packages alphabetically
add code comment

Djfe committed:
site.mk: add code comments from example

Djfe committed:
site.mk: add web-cellular

but commented out

Djfe committed:
site.mk: move some stuff from Makefile to site.mk

remove Release and Checkout stuff from site.mk
set BROKEN in site.mk
set GLUON_AUTOUPDATER_ENABLED in site.mk
set GLUON_OUTPUTDIR in site.mk
set GLUON_PACKAGEDIR in site.mk
move GLUON_DEPRECATED further down in the file

Djfe committed:
Makefile: replace {} by ()

Djfe committed:
Makefile: resort sections

sort targets more chronologically
presort variables for later commits

Djfe committed:
Makefile: Update to new FFMUC changes

create ${GLUON_BUILD_DIR} and fill it without git clone
split first half of gluon-prepare and create gluon-update

we will keep our own gluon-patch for now instead of running their shell script
scripts/apply_patches.sh

Djfe committed:
Makefile: replace " by '

Djfe committed:
Makefile: rework how GLUON_MAKE is called

call GLUON_MAKE as a sub make process
set MAKEFLAGS instead of adding parameters to the command line
stop printing directory changes
synchronize output, useful for parallel execution (jobs) and logging

fix bug with calculation of jobs (available threads to use by make):
- count with grep instead of wc
- only count lines that start with "processor": ^processor
alternatively we could've switched to "nproc" instead

remove GLUON_AUTOUPDATER_BRANCH, default is stable anyways
two variables were move to site.mk by cb208a620c0521703645692d8690d9d5a50b3207
- GLUON_AUTOUPDATER_ENABLED
- BROKEN
remove them
export GLUON_RELEASE
export and set GLUON_SITEDIR for GLUON_MAKE
instead of creating a symbolic link to site dir during gluon_prepare
adjust PATCH_DIR

site.mk sets GLUON_OUTPUTDIR > stop moving files after target manifest
GLUON_MAKE builds right into the output folder in the GLUON_SITEDIR

Djfe committed:
Makefile: redo GLUON_RELEASE, error for missing git tag

Djfe committed:
Makefile: redo info section

list useful information for debugging:
- GLUON_RELEASE, remove autoupdater branch
- TARGETS and DEVICES, if we limit them via command line
- MAKEFLAGS
- git url and ref
- number of patch files

Show info on every execution except of cleaning

Djfe committed:
Makefile: move git refs into release.mk

advantage: makefile stays the same across branches/releases

Djfe committed:
Makefile: rename targets to targets.mk

substitute newlines with spaces in targets.mk

Djfe committed:
Makefile: rename: OPKG_KEY_BUILD_DIR > OPKG_KEY_FOLDER

Djfe committed:
Makefile: explain available Custom Options at the top of the Makefile

Djfe committed:
Makefile: rework gluon-update and gluon-patch

simplify calling git in a subfolder with GLUON_GIT
gluon-update: checkout master, hide&prevent errors when branch empty
list PATCH_FILES with variable for git apply
rename gluon-patch to ffac-patch
hide patch commands from being printed
rework git apply so it runs from sitedir but in a subdirectory

Djfe committed:
Makefile: redo sign, for loop over all manifests

requires SECRET_KEY_FILE to exist to run
control, what is being echoed, hide commands

Djfe committed:
Makefile: manifest explain what's being done (echo)

hide the command that is being executed

Djfe committed:
Makefile: rework gluon-prepare

split gluon-prepare
only update gluon modules when files where changed that require doing so
- patch files
- release.mk
- modules
- the checked out git hash of GLUON_BUILD_DIR changed due to gluon-update

add FORCE target to force comparison of checked out git hash with file on every run

only copy opkg keys to GLUON_BUILD_DIR/openwrt/ when updating gluon modules
> move the line from target build to .modules
only copy opkg keys when they exist in OPKG_KEY_FOLDER
only copy opkg keys when none exist in GLUON_BUILD_DIR/openwrt/

make sure .modules gets rerun, when it's canceled during execution

Djfe committed:
Makefile: site.conf: rework build, handling of opkg packages

hide ugly for loop and if commands

when no keys exist in OPKG_KEY_FOLDER, copy over freshly generated
opkg keys from GLUON_BUILD_DIR/openwrt/ to OPKG_KEY_FOLDER

move opkg packages into folder with OPENWRT_BRANCH in its name
use rsync for copying packages
exclude packages available in openwrt feeds

declare additional gluon opkg feeds in site.conf

Djfe committed:
Makefile: support GLUON_DEVICES better

when it's set:
- disable make sign
- don't move packages to devices folder in build target
- create devices folder during output-clean (build requirement)
- set GLUON_OUTPUTDIR to devices folder

Djfe committed:
Makefile: add a way to update our patches

also delegates GLUON_MAKE to subsequently update Gluon patches
just in case our patch patches OpenWrt or another Gluon module

Djfe committed:
Makefile: disable all implicit rules

also don't check for rebuilding "Makefile" before execution

Djfe committed:
Makefile: add .PHONY target - declare most targets as phony

phony targets are not files and need to be run on every execution
similar to FORCE but newer style
FORCE needs to stay for the git hash target else it won't work (quirk)

Djfe committed:
.gitignore: hide build.log files

Djfe committed:
patches: fix Gluon's Makefile, GLUON_TARGETS specifically

Djfe committed:
run make update-patches

maurerle commented on Makefile:
```diff
@@ -169,11 +166,29 @@ ffac-patch:
 
 clean: gluon-clean output-clean
+
+FORCE: ;
```
what ist this?

Djfe replied:
as explained in the commit message (iirc):
FORCE is the old .PHONY
with the big difference, that FORCE rebuilds everytime

.PHONY appears to have an exception for files when all requirements didn't need an update (rebuild), no clue
.cmp-git-hash only works with a FORCE target

FORCE is just a name everyone uses by convention

.cmp-git-hash is just an implementation of .up-to-date, I got it from here
https://stackoverflow.com/a/24328410
it's all explained here. also why we need FORCE instead of .PHONY

maurerle commented on site.conf:
```diff
@@ -6,8 +6,15 @@
 	domain_seed = 'fcd2bb830c677865c60459899facdb630a9c09e26b72d995451c4feb9b337e6d',
 
 	opkg = {
+		openwrt = 'http://downloads.openwrt.org/releases/packages-%v/%A',
```
is this just being explicit instead of implicit or does this change something?

cbricart replied:
I *think* one has to fully fill the map with the default base (i.e upstream) definition when adding deriving package sources later on (i.e in `extra = { … }`)

Djfe replied:
I haven't had the chance to test this change, yet.
according to readthedocs the default is snapshot not releases/packages-22.03

I didn't want to supply master packets.
run opkg update and upgrade once might actually update tons of packages for which we don't have the available flash.

it's possible that the documentation is wrong, I haven't checked the code

> openwrt overrides the default OpenWrt repository URL. The default URL would correspond to http://downloads.openwrt.org/snapshots/packages/%A and usually doesn’t need to be changed when nodes are expected to have IPv6 internet connectivity.

https://gluon.readthedocs.io/en/latest/user/site.html

@cbricart we could also use your mirror of openwrt here if it improves anything.
if it's just a forward then nvm
http://opkg.ffac.rocks/ (< read)

cbricart replied:
we (have) had the full mirror [actually reverse-caching-proxy to be precise :)] in place when upstream openwrt was IPv4 only. This was changed back to use upstream servers again when they switched on being dual stack.

Djfe replied:
actually I got the openwrt URL from Freifunk Darmstadt or Hannover
https://github.com/freifunk-darmstadt/site-ffda/blob/master/site.conf#L23
https://github.com/freifunkh/site/blob/aba526211f7d1e49259c29dc701ba2a072e3db1d/site.conf#L9

Djfe replied:
maybe it's actually possible to combine all of these feeds into one link

maybe I need a feeds.conf, like here? (maybe I don't and it just works)
https://downloads.openwrt.org/releases/packages-22.03/mips_24kc/

Djfe replied:
Update:
```
/etc/opkg# cat distfeeds.conf 
# src/gz openwrt_core https://downloads.openwrt.org/releases/22.03-SNAPSHOT/targets/ath79/generic/packages
src/gz openwrt_base https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/mips_24kc/base
src/gz openwrt_luci https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/mips_24kc/luci
src/gz openwrt_packages https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/mips_24kc/packages
src/gz openwrt_routing https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/mips_24kc/routing
src/gz openwrt_telephony https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/mips_24kc/telephony
```
This is how the OpenWrt feeds are setup, when we don't change anything.
If you take a look at:
https://downloads.openwrt.org/releases/22.03-SNAPSHOT, then you'll see, that packages is just a symlink
[packages](https://downloads.openwrt.org/releases/22.03-SNAPSHOT/packages/)/ -> ../packages-22.03

therefore changing the path above doesn't change a thing, actually
I'll therefore remove the line specifying openwrt again from our site.conf

furthermore:
I realized that the packages from the feeds base, packages and routing are just subsets from openwrt, therefore I'll remove those lines as well.
Which leaves: modules (our version of openwrt_core, which is commented out in distfeeds.conf for a reason)
and ffac, gluon and gluon_base

Those will all be defined/listed in /etc/opkg/gluon.conf (right now it contains only modules)

is mistook snapshots for master-based packages btw.
while it actually meant snapshots of the 22.03 branch, which is fine

Djfe commented:
I recoded how the GLUON_RELEASE is set
I fixed a few things with commit messages
I renamed "our" to "your" (in target .modules) and added another echo message to target "build" (both changes are in regards to opkg keys being copied)
I moved two strings further down in the Makefile (in regards to targets and devices and the info section)
The Custom Options explain now, that GLUON_DEBUG needs a `make clean` if you care for all packages to be rebuilt with debug symbols.

last thing to test now is opkg stuff, I'm doing that together with @cbricart

Djfe commented:
opkg packages are now copied by rsync
mostly rsync instead of cp, so I can exclude some folders

Djfe commented:
I just renamed a patch so the name uses "-" instead of "_"

Djfe committed:
redo gluon-packages.patch so it can be merged into other branches

it's now a patch files that gluon applies
this removes the requirement to modify all branches individually
just for adding a script that is run on update

Djfe committed:
Makefile: fix update-patches for module patches

$(GLUON_MAKE) update-patches runs reliably now.
Beforehand it created artifacts if there were
module patches

Djfe committed:
Makefile: create & edit patches

Djfe committed:
site.mk remove duplicate line
